File: java/integration-tests/src/test/java/io/joynr/integration/MultipleVersionsTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2018 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.integration;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.Properties;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import com.google.inject.Module;
import com.google.inject.util.Modules;

import io.joynr.arbitration.DiscoveryQos;
import io.joynr.arbitration.DiscoveryScope;
import io.joynr.exceptions.DiscoveryException;
import io.joynr.exceptions.JoynrRuntimeException;
import io.joynr.integration.util.DummyJoynrApplication;
import io.joynr.messaging.MessagingPropertyKeys;
import io.joynr.messaging.routing.TestGlobalAddressModule;
import io.joynr.provider.AbstractJoynrProvider;
import io.joynr.proxy.Future;
import io.joynr.proxy.ProxyBuilder;
import io.joynr.proxy.ProxyBuilder.ProxyCreatedCallback;
import io.joynr.runtime.CCInProcessRuntimeModule;
import io.joynr.runtime.JoynrInjectorFactory;
import io.joynr.runtime.JoynrRuntime;
import joynr.tests.DefaultMultipleVersionsInterface1Provider;
import joynr.tests.DefaultMultipleVersionsInterface2Provider;
import joynr.tests.DefaultMultipleVersionsInterfaceProvider;
import joynr.tests.MultipleVersionsInterface1Proxy;
import joynr.tests.MultipleVersionsInterface2Proxy;
import joynr.tests.v2.MultipleVersionsInterfaceProxy;
import joynr.types.ProviderQos;
import joynr.types.ProviderScope;

public class MultipleVersionsTest {
    private static final String DOMAIN = "MultipleVersionsTestDomain";
    private static final String PROXYBUILD_FAILED_MESSAGE = "Building of proxy failed: ";
    private static final String REGISTERING_FAILED_MESSAGE = "Registering of provider failed: ";
    private static final String UNREGISTERING_FAILED_MESSAGE = "Unregistering of provider failed: ";

    private Semaphore proxyBuilt;

    private DiscoveryQos discoveryQos;
    private ProviderQos providerQos;

    private JoynrRuntime runtime;

    @Before
    public void setUp() {
        discoveryQos = new DiscoveryQos();
        discoveryQos.setDiscoveryTimeoutMs(5000);
        discoveryQos.setRetryIntervalMs(100);
        discoveryQos.setDiscoveryScope(DiscoveryScope.LOCAL_ONLY);

        providerQos = new ProviderQos();
        providerQos.setScope(ProviderScope.LOCAL);

        proxyBuilt = new Semaphore(0);

        Properties joynrConfig = new Properties();
        joynrConfig.setProperty(MessagingPropertyKeys.CHANNELID, "discoverydirectory_channelid");

        Module runtimeModule = Modules.override(new CCInProcessRuntimeModule()).with(new TestGlobalAddressModule());
        DummyJoynrApplication application = (DummyJoynrApplication) new JoynrInjectorFactory(joynrConfig,
                                                                                             runtimeModule).createApplication(DummyJoynrApplication.class);

        runtime = application.getRuntime();
    }

    @After
    public void tearDown() {
        runtime.shutdown(true);
    }

    /**
     * Registers given provider in runtime.
     * @param provider
     */
    private void asyncRegisterProvider(AbstractJoynrProvider provider) {
        Future<Void> future = runtime.registerProvider(DOMAIN, provider, providerQos);

        try {
            future.get(100);
        } catch (Exception e) {
            fail(REGISTERING_FAILED_MESSAGE + e);
        }
    }

    /**
     * Builds a proxy of the requested interface in runtime.
     * @param interfaceClass Interface of the proxy to build.
     * @return The built proxy.
     */
    private <T> T asyncBuildProxyForInterface(final Class<T> interfaceClass) {
        ProxyBuilder<T> proxyBuilder = runtime.getProxyBuilder(DOMAIN, interfaceClass);
        T proxy = null;
        try {
            proxy = proxyBuilder.setDiscoveryQos(discoveryQos).build(new ProxyCreatedCallback<T>() {
                @Override
                public void onProxyCreationFinished(T result) {
                    proxyBuilt.release();
                }

                @Override
                public void onProxyCreationError(JoynrRuntimeException error) {
                    throw error;
                }
            });
            assertTrue(proxyBuilt.tryAcquire(1, TimeUnit.SECONDS));
        } catch (DiscoveryException | InterruptedException e) {
            fail(PROXYBUILD_FAILED_MESSAGE + e);
        }
        return proxy;
    }

    /*
    * This test tests if 2 proxies of same interface (name versioned and package versioned) can connect to
    * one provider (unversioned) and communicate with this without mutual interference.
    */
    @Test
    public void twoProxiesOfDifferentVersioningTypesVsUnversionedProvider() {
        // register provider
        DefaultMultipleVersionsInterfaceProvider provider = new DefaultMultipleVersionsInterfaceProvider();

        asyncRegisterProvider(provider);

        // build fitting proxies
        MultipleVersionsInterface2Proxy proxy1 = asyncBuildProxyForInterface(MultipleVersionsInterface2Proxy.class);
        MultipleVersionsInterfaceProxy proxy2 = asyncBuildProxyForInterface(MultipleVersionsInterfaceProxy.class);

        try {
            //set UInt8Attribute1 and check if it can be retrieved correctly
            proxy1.setUInt8Attribute1((byte) 100);
            Byte value1 = proxy1.getUInt8Attribute1();
            Byte value2 = proxy2.getUInt8Attribute1();
            assertEquals((byte) value1, 100);
            assertEquals((byte) value2, 100);

            proxy2.setUInt8Attribute1((byte) 50);
            value1 = proxy1.getUInt8Attribute1();
            value2 = proxy2.getUInt8Attribute1();
            assertEquals((byte) value1, 50);
            assertEquals((byte) value2, 50);

            // unregister provider
            runtime.unregisterProvider(DOMAIN, provider);
        } catch (JoynrRuntimeException e) {
            fail(UNREGISTERING_FAILED_MESSAGE + e);
        }
    }

    /*
     * This test tests if 2 providers of same interface and different versions can be registered in a single runtime
     * and 2 proxies can communicate with those without mutual interference.
     */
    @Test
    @Ignore
    public void twoNameVersionedProvidersInSingleRuntime() {
        // register providers
        DefaultMultipleVersionsInterface1Provider provider1 = new DefaultMultipleVersionsInterface1Provider();
        DefaultMultipleVersionsInterface2Provider provider2 = new DefaultMultipleVersionsInterface2Provider();

        asyncRegisterProvider(provider1);
        asyncRegisterProvider(provider2);

        // build fitting proxies
        MultipleVersionsInterface1Proxy proxy1 = asyncBuildProxyForInterface(MultipleVersionsInterface1Proxy.class);
        MultipleVersionsInterface2Proxy proxy2 = asyncBuildProxyForInterface(MultipleVersionsInterface2Proxy.class);

        //set UInt8Attribute1 and check if it can be retrieved correctly
        proxy1.setUInt8Attribute1((byte) 100);
        proxy2.setUInt8Attribute1((byte) 50);
        Byte value1 = proxy1.getUInt8Attribute1();
        Byte value2 = proxy2.getUInt8Attribute1();
        assertEquals((byte) value1, 100);
        assertEquals((byte) value2, 50);

        // unregister providers
        try {
            runtime.unregisterProvider(DOMAIN, provider1);
            runtime.unregisterProvider(DOMAIN, provider2);
        } catch (JoynrRuntimeException e) {
            fail(UNREGISTERING_FAILED_MESSAGE + e);
        }
    }

}

File: javascript/libjoynr-js/src/test/js/node_integration/MultipleVersionsTest.js
/*eslint-disable */

/*
 * #%L
 * %%
 * Copyright (C) 2018 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
let IntegrationUtils = require("./IntegrationUtils");
let MultipleVersionsInterfaceProviderNameVersion1 = require("../../generated/joynr/tests/MultipleVersionsInterface1Provider");
let MultipleVersionsInterfaceProviderNameVersion2 = require("../../generated/joynr/tests/MultipleVersionsInterface2Provider");
let MultipleVersionsInterfaceProviderPackageVersion1 = require("../../generated/joynr/tests/v1/MultipleVersionsInterfaceProvider");
let MultipleVersionsInterfaceProviderPackageVersion2 = require("../../generated/joynr/tests/v2/MultipleVersionsInterfaceProvider");
let MultipleVersionsInterfaceProviderUnversioned = require("../../generated/joynr/tests/MultipleVersionsInterfaceProvider");
let MultipleVersionsInterfaceProxyNameVersion1 = require("../../generated/joynr/tests/MultipleVersionsInterface1Proxy");
let MultipleVersionsInterfaceProxyNameVersion2 = require("../../generated/joynr/tests/MultipleVersionsInterface2Proxy");
let MultipleVersionsInterfaceProxyPackageVersion2 = require("../../generated/joynr/tests/v2/MultipleVersionsInterfaceProxy");
let providerImplementation = require("./MultipleVersionsInterfaceProviderImplementation");
let provisioning = require("../../resources/joynr/provisioning/provisioning_cc.js");

const domain = "MultipleVersionsTestDomain";
let joynr = require("joynr");

describe("libjoynr-js.integration.MultipleVersionsTest", () => {
    beforeEach(async () => {
        joynr.loaded = false;
        joynr.selectRuntime("inprocess");
        await joynr.load(provisioning);
    });

    afterEach(async () => {
        await joynr.shutdown();

        delete require.cache;
        // remove old joynr exit handler
        process.removeAllListeners("exit");

        // rerequire moduls
        MultipleVersionsInterfaceProviderNameVersion1 = require("../../generated/joynr/tests/MultipleVersionsInterface1Provider");
        MultipleVersionsInterfaceProviderNameVersion2 = require("../../generated/joynr/tests/MultipleVersionsInterface2Provider");
        MultipleVersionsInterfaceProviderPackageVersion1 = require("../../generated/joynr/tests/v1/MultipleVersionsInterfaceProvider");
        MultipleVersionsInterfaceProviderPackageVersion2 = require("../../generated/joynr/tests/v2/MultipleVersionsInterfaceProvider");
        MultipleVersionsInterfaceProviderUnversioned = require("../../generated/joynr/tests/MultipleVersionsInterfaceProvider");
        MultipleVersionsInterfaceProxyNameVersion1 = require("../../generated/joynr/tests/MultipleVersionsInterface1Proxy");
        MultipleVersionsInterfaceProxyNameVersion2 = require("../../generated/joynr/tests/MultipleVersionsInterface2Proxy");
        MultipleVersionsInterfaceProxyPackageVersion2 = require("../../generated/joynr/tests/v2/MultipleVersionsInterfaceProxy");
        joynr = require("joynr");
    });

    async function buildProvider(providerType) {
        const providerQos = new joynr.types.ProviderQos({
            customParameters: [],
            priority: 5,
            scope: joynr.types.ProviderScope.LOCAL,
            supportsOnChangeSubscriptions: false
        });
        const multipleVersionsInterfaceProvider = joynr.providerBuilder.build(providerType, providerImplementation);

        await joynr.registration.registerProvider(domain, multipleVersionsInterfaceProvider, providerQos);

        return multipleVersionsInterfaceProvider;
    }

    async function buildProxy(proxyType) {
        const discoveryQos = new joynr.proxy.DiscoveryQos({
            discoveryTimeoutMs: 1000
        });
        return await joynr.proxyBuilder.build(proxyType, {
            domain,
            discoveryQos
        });
    }

    // set same attribute of different providers and check if right result is obtained.
    async function setUInt8AttributesAndCheck(proxy1, proxy2) {
        // set attributes
        await Promise.all([proxy1.uInt8Attribute1.set({ value: 50 }), proxy2.uInt8Attribute1.set({ value: 100 })]);

        // retreive attributes
        const [value1, value2] = await Promise.all([proxy1.uInt8Attribute1.get(), proxy2.uInt8Attribute1.get()]);

        // check if received values equal set values.
        expect(value1).toEqual(50);
        expect(value2).toEqual(100);
    }

    /**
     * Two providers of different version can not be registered in a single runtime.
     * (They are both registered with the same participant id and thus seen as one.)
     * @Todo: Get this test to pass.
     */
    xit("2 proxies and 2 providers of different versions in same runtime (name version vs. name version)", async () => {
        // build and register providers
        const [provider1, provider2] = await Promise.all([
            buildProvider(MultipleVersionsInterfaceProviderNameVersion1),
            buildProvider(MultipleVersionsInterfaceProviderNameVersion2)
        ]);

        // build fitting proxies
        const [proxy1, proxy2] = await Promise.all([
            buildProxy(MultipleVersionsInterfaceProxyNameVersion1),
            buildProxy(MultipleVersionsInterfaceProxyNameVersion2)
        ]);

        // set primitive type attributes and check if values can be retrieved properly by proxies
        // without mutual interference
        await setUInt8AttributesAndCheck(proxy1, proxy2);

        // unregister providers
        await Promise.all([
            joynr.registration.unregisterProvider(domain, provider1),
            joynr.registration.unregisterProvider(domain, provider2)
        ]);
    });

    /**
     * Two providers of different version can not be registered in a single runtime.
     * (They are both registered with the same participant id and thus seen as one.)
     * @Todo: Get this test to pass.
     */
    xit("2 proxies and 2 providers of different versions in same runtime (name version vs. package version)", async () => {
        // build and register providers
        const [provider1, provider2] = await Promise.all([
            buildProvider(MultipleVersionsInterfaceProviderPackageVersion1),
            buildProvider(MultipleVersionsInterfaceProviderPackageVersion2)
        ]);

        // build fitting proxies
        const [proxy1, proxy2] = await Promise.all([
            buildProxy(MultipleVersionsInterfaceProxyNameVersion1),
            buildProxy(MultipleVersionsInterfaceProxyNameVersion2)
        ]);

        // set primitive type attributes and check if values can be retrieved properly by proxies
        // without mutual interference
        await setUInt8AttributesAndCheck(proxy1, proxy2);

        // unregister providers
        await Promise.all([
            joynr.registration.unregisterProvider(domain, provider1),
            joynr.registration.unregisterProvider(domain, provider2)
        ]);
    });

    it("2 proxies of different versioning types vs. unversioned provider", async () => {
        // build and register provider
        const multipleVersionsInterfaceProvider1 = await buildProvider(MultipleVersionsInterfaceProviderUnversioned);
        // build fitting proxies
        const [proxy1, proxy2] = await Promise.all([
            buildProxy(MultipleVersionsInterfaceProxyNameVersion2),
            buildProxy(MultipleVersionsInterfaceProxyPackageVersion2)
        ]);

        // check if attributes can be set and retrieved correctly by both proxies
        await Promise.all([proxy1.uInt8Attribute2.set({ value: 100 }), proxy2.uInt8Attribute1.set({ value: 50 })]);

        const [value1, value2] = await Promise.all([proxy1.uInt8Attribute1.get(), proxy2.uInt8Attribute2.get()]);

        expect(value1).toEqual(50);
        expect(value2).toEqual(100);

        await joynr.registration.unregisterProvider(domain, multipleVersionsInterfaceProvider1);
    });

    it("2 proxies connecting to 2 providers in different ChildProcesses subscribing to the same type (non primitive)", async () => {
        // register 2 providers in 2 child processes
        const [childId1, childId2] = await Promise.all([
            IntegrationUtils.initializeChildProcess("TestMultipleVersionsInterfaceProcess", "", domain, {
                versioning: "nameVersion1",
                noJoynrShutdown: true
            }),
            IntegrationUtils.initializeChildProcess("TestMultipleVersionsInterfaceProcess", "", domain, {
                versioning: "nameVersion2",
                noJoynrShutdown: true
            })
        ]);

        // build fitting proxies
        const [proxy1, proxy2] = await Promise.all([
            buildProxy(MultipleVersionsInterfaceProxyNameVersion1),
            buildProxy(MultipleVersionsInterfaceProxyNameVersion2)
        ]);

        // set attributes and check if values can be retrieved properly by proxies without mutual interference
        await setUInt8AttributesAndCheck(proxy1, proxy2);

        // unregister providers and kill child processes
        await Promise.all([
            IntegrationUtils.shutdownChildProcess(childId2),
            IntegrationUtils.shutdownChildProcess(childId1)
        ]);
    });
});
/*eslint-enable */
